src/full/Agda/TypeChecking/Conversion.hs

```diff
@@ -1761,24 +1763,38 @@ equalSort s1 s2 = do
 
       -- Equate a sort @s1@ to @univSort s2@
       -- Precondition: @s1@ and @univSort s2@ are already reduced.
-      univSortEquals :: Bool -> Sort -> Sort -> Blocker -> m ()
-      univSortEquals propEnabled s1 s2 blocker = do
+      univSortEquals :: Bool -> Bool -> Sort -> Sort -> Blocker -> m ()
+      univSortEquals propEnabled infInInf s1 s2 blocker = do
         reportSDoc "tc.conv.sort" 35 $ vcat
           [ "univSortEquals"
           , "  s1 =" <+> prettyTCM s1
           , "  s2 =" <+> prettyTCM s2
           ]
         let postpone = patternViolation blocker
         case s1 of
+          -- @Prop l@, @SizeUniv@ and @LevelUniv@ are not successor sorts.
+          Prop{}      -> no
+          Inf UProp _ -> no
+          SizeUniv{}  -> no
+          LevelUniv{} -> no
+          -- Neither are @LockUniv@ or @IntervalUniv@.
+          LockUniv{}     -> no
+          IntervalUniv{} -> no
+
           -- @Set l1@ is the successor sort of either @Set l2@ or
           -- @Prop l2@ where @l1 == lsuc l2@.
           Type l1 -> do
+            levelUnivEnabled <- optLevelUniverse <$> pragmaOptions
+            guardedEnabled   <- optGuarded       <$> pragmaOptions
                -- @s2@ is definitely not @Inf n@ or @SizeUniv@
-            if | Inf _ n  <- s2 -> no
-               | SizeUniv <- s2 -> no
-               -- If @Prop@ is not used, then @s2@ must be of the form
-               -- @Set l2@
-               | not propEnabled -> do
+            if | Inf _ _n <- s2 -> __IMPOSSIBLE__
+               | SizeUniv <- s2 -> __IMPOSSIBLE__
+               -- The predecessor @s2@ is can also not be @SSet _@ or @IntervalUniv@
+               | Univ USSet _ <- s2 -> __IMPOSSIBLE__
+               | IntervalUniv <- s2 -> __IMPOSSIBLE__
```

Comment:
@jespercockx These cases should all be impossible, so maybe we want to remove them altogether... WDYT?

Reply:
Well if they are not too expensive we could leave them in as assertions that check that we do not violate this precondition.

Reply:
Ok!